### DISCUSSION TITLE
Duplicate model schemas are generated when enum is used in nested model

### DISCUSSION THREAD

**Discussion by hekrause** [fastapi/fastapi#11030] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from pydantic import BaseModel
from enum import Enum

app = FastAPI()


class StatusEnum(str, Enum):
    STATUS_A = "status_a"
    STATUS_B = "status_b"


class NestedModelWithEnum(BaseModel):
    name: str
    status: StatusEnum


class NestedModelWithoutEnum(BaseModel):
    name: str
    status: str


class TopModelA(BaseModel):
    name: str
    nested_model: NestedModelWithEnum


class TopModelB(BaseModel):
    name: str
    nested_model: NestedModelWithEnum


class TopModelC(BaseModel):
    name: str
    nested_model: NestedModelWithEnum


class TopModelZ(BaseModel):
    name: str
    nested_model: NestedModelWithoutEnum


@app.get("/A_input_and_output")
def read_root1(model: TopModelA) -> TopModelA:
    return TopModelA()


@app.get("/Z_input_and_output")
def read_root2(model: TopModelZ) -> TopModelZ:
    return TopModelZ()


@app.get("/Nested_input_and_output")
def read_root3(model: NestedModelWithEnum) -> NestedModelWithEnum:
    return NestedModelWithEnum()


@app.get("/B_output")
def read_root4() -> TopModelB:
    return TopModelB()


@app.get("/C_input")
def read_root5(model: TopModelC) -> bool:
    return True
```

### Description

Hi all,

i encountered a weird problem and reduced it down to a minimal example.

When i define endpoints and start the fastapi with uvicorn the generated openapi spec has double schema entries for certain models. These schema entries are named like MODELNAME-Input and MODELNAME-Output although the model is named MODELNAME in my code. To further use the openapi-spec definiton this is very ugly and i would say this is undefined or unwanted behaviour of fastapi or openapi because you never now how the schema will be called. Also the "different" schemas are not different because they have the exact same definition in the openapi.json. I use the openapi.json to generate a frontend javascript client and there this causes the hick up.

The error only occurs when the model MODELNAME is used as request parameter and also as response parameter in endpoints AND when this model has a nested model with an enum in it.

![image](https://github.com/tiangolo/fastapi/assets/17926729/4480d76d-5dfe-496f-8da9-3fa44dea67c1)

As you can see in my example code the endpoint "/A_input_and_output" receive a parameter TopModelA and return an object of TopModelA. In the image you can see the new schema entries for TopModelA-Input and TopModelA-Output.

The endpoint "/Z_input_and_output" uses the same model TopModelZ as request and response type like the endpoint "/A_input_and_output" but there are no duplicate schemas generated.

The TopModelB and TopModelC are either used as Request or as Response parameter but not both so there the problem does not occure.

All three models TopModelA, TopModelB and TopModelC consists of a field and an nested model and in this model is an enum.
The model TopModelZ however has also a nested model but no enum in this.

So i think there is some kind of equals operation between the input and output schemas that does not work well with an enum in a nested model. The enum itself does not seem to cause the problem because the endpoint "/Nested_input_and_output" works well with this.

I open the discussion/issue here and not at pydantic or openapi/swagger.io because i can not tell where to locate the problem in this kind of stack.

Many thanks for reading this long text.

openapi.json:

{"openapi":"3.1.0","info":{"title":"FastAPI","version":"0.1.0"},"paths":{"/A_input_and_output":{"get":{"summary":"Read Root1","operationId":"read_root1_A_input_and_output_get","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelA-Input"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelA-Output"}}}},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}}}},"/Z_input_and_output":{"get":{"summary":"Read Root2","operationId":"read_root2_Z_input_and_output_get","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelZ"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelZ"}}}},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}}}},"/Nested_input_and_output":{"get":{"summary":"Read Root3","operationId":"read_root3_Nested_input_and_output_get","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/NestedModelWithEnum"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"$ref":"#/components/schemas/NestedModelWithEnum"}}}},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}}}},"/B_output":{"get":{"summary":"Read Root4","operationId":"read_root4_B_output_get","responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelB"}}}}}}},"/C_input":{"get":{"summary":"Read Root5","operationId":"read_root5_C_input_get","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/TopModelC"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"type":"boolean","title":"Response Read Root5 C Input Get"}}}},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}}}}},"components":{"schemas":{"HTTPValidationError":{"properties":{"detail":{"items":{"$ref":"#/components/schemas/ValidationError"},"type":"array","title":"Detail"}},"type":"object","title":"HTTPValidationError"},"NestedModelWithEnum":{"properties":{"name":{"type":"string","title":"Name"},"status":{"$ref":"#/components/schemas/StatusEnum"}},"type":"object","required":["name","status"],"title":"NestedModelWithEnum"},"NestedModelWithoutEnum":{"properties":{"name":{"type":"string","title":"Name"},"status":{"type":"string","title":"Status"}},"type":"object","required":["name","status"],"title":"NestedModelWithoutEnum"},"StatusEnum":{"type":"string","enum":["status_a","status_b"],"title":"StatusEnum"},"TopModelA-Input":{"properties":{"name":{"type":"string","title":"Name"},"nested_model":{"$ref":"#/components/schemas/NestedModelWithEnum"}},"type":"object","required":["name","nested_model"],"title":"TopModelA"},"TopModelA-Output":{"properties":{"name":{"type":"string","title":"Name"},"nested_model":{"$ref":"#/components/schemas/NestedModelWithEnum"}},"type":"object","required":["name","nested_model"],"title":"TopModelA"},"TopModelB":{"properties":{"name":{"type":"string","title":"Name"},"nested_model":{"$ref":"#/components/schemas/NestedModelWithEnum"}},"type":"object","required":["name","nested_model"],"title":"TopModelB"},"TopModelC":{"properties":{"name":{"type":"string","title":"Name"},"nested_model":{"$ref":"#/components/schemas/NestedModelWithEnum"}},"type":"object","required":["name","nested_model"],"title":"TopModelC"},"TopModelZ":{"properties":{"name":{"type":"string","title":"Name"},"nested_model":{"$ref":"#/components/schemas/NestedModelWithoutEnum"}},"type":"object","required":["name","nested_model"],"title":"TopModelZ"},"ValidationError":{"properties":{"loc":{"items":{"anyOf":[{"type":"string"},{"type":"integer"}]},"type":"array","title":"Location"},"msg":{"type":"string","title":"Message"},"type":{"type":"string","title":"Error Type"}},"type":"object","required":["loc","msg","type"],"title":"ValidationError"}}}}

### Operating System

Windows

### Operating System Details

=> Pycharm Professional IDE
=> start of program from powershell: uvicorn app:app --host 0.0.0.0 --port 8002
=> visual inspection of openapi spec via firefox at http://localhost:8002/docs and http://localhost:8002/openapi.json

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.10.11

### Additional Context

requirements.txt
===============
annotated-types==0.6.0
anyio==4.2.0
click==8.1.7
colorama==0.4.6
exceptiongroup==1.2.0
fastapi==0.109.0
h11==0.14.0
idna==3.6
pydantic==2.5.3
pydantic_core==2.14.6
sniffio==1.3.0
starlette==0.35.1
typing_extensions==4.9.0
uvicorn==0.27.0

**Comment by niqzart** [comment]
Hello! I've had a bit of a different issue with duplication, and this is one of the first links in google, so I'm leaving a breadcrumb for future devs with the same issue

Maybe the answers in [this discussion](https://github.com/fastapi/fastapi/discussions/10019) ([this comment in particular](https://github.com/fastapi/fastapi/discussions/10019#discussioncomment-6826463)) can help

Although, it is strange, that models with no defaults are treated this way, maybe that's a bug
